core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/PromiseCombiner.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.util.concurrent;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import net.jcip.annotations.ThreadSafe;
+
+/**
+ * A thread-safe version of Netty's {@link io.netty.util.concurrent.PromiseCombiner} that uses
+ * proper synchronization to trigger the completion of the aggregate promise.
+ */
+@ThreadSafe
+public class PromiseCombiner {
+
+  /**
+   * Combines the given futures into the given promise, that is, ties the completion of the latter
+   * to that of the formers.
+   *
+   * @param aggregatePromise The promise that will complete when all parents complete.
+   * @param parents The parent futures.
+   */
+  public static void combine(
+      @NonNull Promise<Void> aggregatePromise, @NonNull Future<?>... parents) {
+    PromiseCombinerListener listener =
+        new PromiseCombinerListener(aggregatePromise, parents.length);
+    for (Future<?> parent : parents) {
+      parent.addListener(listener);
+    }
+  }
+
+  private static class PromiseCombinerListener implements GenericFutureListener<Future<Object>> {
+
+    private final Promise<Void> aggregatePromise;
+    private final AtomicInteger remainingCount;
+    private final AtomicReference<Throwable> aggregateFailureRef = new AtomicReference<>();
+
+    private PromiseCombinerListener(Promise<Void> aggregatePromise, int numberOfParents) {
+      this.aggregatePromise = aggregatePromise;
+      remainingCount = new AtomicInteger(numberOfParents);
+    }
+
+    @Override
+    public void operationComplete(Future<Object> future) {
+      if (!future.isSuccess()) {
+        aggregateFailureRef.updateAndGet(
+            aggregateFailure -> {
+              if (aggregateFailure == null) {
+                aggregateFailure = future.cause();
+              } else {
+                aggregateFailure.addSuppressed(future.cause());
```

Comment:
is it possible that `future.cause()` will be `null`? or `future`? if so our logic may not work because of this check: https://github.com/datastax/java-driver/pull/1371/files/8ae9ee79a503454657dda8814ccb3b132e6e0db4..3242753311848dd161315230777949c7de750dc3#diff-ab0e22e2e333da4968ffb1f8318c49afR75

Reply:
`future` itself cannot be null. `future.cause()` can be null if 1) the future is not completed yet, or if it completed successfully. But here these conditions are not met, since when `operationComplete` is called the future is necessarily complete, and since we tested that `future.isSuccess() == false`, we know that the future failed.